fix: (storage) bucket policy delete 412 error

gurusai-voleti opened the pull request:
<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/20838
**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:bug
storage: fixed a 412 error returned on some `google_storage_bucket_iam_policy` deletions
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 9 insertions(+), 868 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 9 insertions(+), 868 deletions(-))

## Breaking Change(s) Detected

The following breaking change(s) were detected within your pull request.
- Resource `google_storage_bucket_iam_binding` was either removed or renamed - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#resource-map-resource-removal-or-rename)
- Resource `google_storage_bucket_iam_member` was either removed or renamed - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#resource-map-resource-removal-or-rename)
- Resource `google_storage_bucket_iam_policy` was either removed or renamed - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#resource-map-resource-removal-or-rename)

If you believe this detection to be incorrect please raise the concern with your reviewer.
If you intend to make this change you will need to wait for a [major release](https://www.terraform.io/plugin/sdkv2/best-practices/versioning#example-major-number-increments) window.
An `override-breaking-change` label can be added to allow merging.

modular-magician commented:
#### Tests analytics
Total tests: 109
Passed tests: 101
Skipped tests: 7
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccStorageBucketIamPolicy</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccStorageBucketIamPolicy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/43192e31-d9df-4550-abfd-45ca3f8fc5f2/build-log/recording_build/TestAccStorageBucketIamPolicy_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/43192e31-d9df-4550-abfd-45ca3f8fc5f2/recording/TestAccStorageBucketIamPolicy.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/43192e31-d9df-4550-abfd-45ca3f8fc5f2/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/43192e31-d9df-4550-abfd-45ca3f8fc5f2/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 13 insertions(+), 868 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 13 insertions(+), 868 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 109
Passed tests: 101
Skipped tests: 7
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccStorageBucketIamPolicy</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccStorageBucketIamPolicy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/90863d23-6e9d-46a4-a323-9df2c293360d/recording/TestAccStorageBucketIamPolicy.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/90863d23-6e9d-46a4-a323-9df2c293360d/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/90863d23-6e9d-46a4-a323-9df2c293360d/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 13 insertions(+), 883 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 13 insertions(+), 883 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 109
Passed tests: 102
Skipped tests: 7
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/0ad6d5ad-1b9d-4a74-9493-e4771c0e8e60/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 110
Passed tests: 103
Skipped tests: 7
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/2845bac6-18a2-4c7c-af98-b7e62346e410/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 110
Passed tests: 103
Skipped tests: 7
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/fb503169-47d6-4178-8aeb-dac2d48a1a26/build-log/replaying_test.log)

rileykarson reviewed (commented):
It seems like we added the etag on delete for GCS bucket in the first place. I can't find context on why yet, but looking into it: https://github.com/GoogleCloudPlatform/magic-modules/pull/2794

edit: original context in https://github.com/hashicorp/terraform-provider-google/pull/1190#issuecomment-377250003

When a blank `bindings` block is sent we get back:

```
{
  "error": {
    "code": 400,
    "message": "A policy to update must be provided.",
    "errors": [
      {
        "message": "A policy to update must be provided.",
        "domain": "global",
        "reason": "required"
      }
    ]
  }
}
```

gurusai-voleti commented:
> It seems like we added the etag on delete for GCS bucket in the first place. I can't find context on why yet, but looking into it: #2794
> 
> edit: original context in [hashicorp/terraform-provider-google#1190 (comment)](https://github.com/hashicorp/terraform-provider-google/pull/1190#issuecomment-377250003)
> 
> When just an empty `bindings` block is sent we get back:
> 
> ```
> {
>   "error": {
>     "code": 400,
>     "message": "A policy to update must be provided.",
>     "errors": [
>       {
>         "message": "A policy to update must be provided.",
>         "domain": "global",
>         "reason": "required"
>       }
>     ]
>   }
> }
> ```

agree but the request body we are sending will look like below
{
"version": 3
}

the request is accepted and policy got updated and response is 200

kautikdk reviewed (commented):
I believe there is a server side validation for rejecting an empty `Buckets:setIamPolicy` request. I tested compute `setIamPolicy` method which allows empty policy and clears if provided. Specifying policy version in the Buckets:setIamPolicy method request should be enough to clear IAM bindings.

rileykarson commented:
Yep- in my testing any value in addition to the empty `bindings` array passed the validation and cleared the bindings. Sorry for the delay on my end- considering whether we can just remove the etag value here for all IAM policies or not, since that would obviate this change.

gurusai-voleti commented:
> Yep- in my testing any value in addition to the empty `bindings` array passed the validation and cleared the bindings. Sorry for the delay on my end- considering whether we can just remove the etag value here for all IAM policies or not, since that would obviate this change.

I added below logic to remove etag only when bindings are empty or null in this change and converted auto generated to handwritten resource, below is only extra logic added to existing iam_bucket

if policy.Bindings == nil || len(policy.Bindings) == 0 {
		policy.Etag = ""
}

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 110
Passed tests: 103
Skipped tests: 7
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/2e5027a9-a672-4783-9cdf-099c4988ca23/build-log/replaying_test.log)

kautikdk reviewed (commented):
Do update here when PR is ready for review. I don't see handwritten documentation file and tests for other cases.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 6 files changed, 99 insertions(+), 883 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 110
Passed tests: 103
Skipped tests: 7
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/750dc15c-7e87-4cf2-8972-0dbd64dadf9c/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 123 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 123 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 118
Passed tests: 103
Skipped tests: 9
Affected tests: 6

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccStorageBucket_iamBindingWithAndWithoutCondition
- TestAccStorageBucket_iamMemberWithAndWithoutCondition
#### Action taken
<details>
<summary>Found 6 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccStorageBucket_basicIamBinding</li>
<li>TestAccStorageBucket_basicIamMemeber</li>
<li>TestAccStorageBucket_basicIamPolicy</li>
<li>TestAccStorageBucket_iamBindingWithCondition</li>
<li>TestAccStorageBucket_iamMemberWithCondition</li>
<li>TestAccStorageBucket_iamPolicyGeneratedWithCondition</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccStorageBucket_basicIamBinding` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_basicIamBinding.log)]
`TestAccStorageBucket_basicIamMemeber` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_basicIamMemeber.log)]
`TestAccStorageBucket_basicIamPolicy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_basicIamPolicy.log)]
`TestAccStorageBucket_iamBindingWithCondition` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_iamBindingWithCondition.log)]
`TestAccStorageBucket_iamMemberWithCondition` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_iamMemberWithCondition.log)]
`TestAccStorageBucket_iamPolicyGeneratedWithCondition` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording/TestAccStorageBucket_iamPolicyGeneratedWithCondition.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/18286a10-ec9a-4ed1-abff-0c90f9a77d89/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 123 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 123 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 118
Passed tests: 109
Skipped tests: 9
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccStorageBucket_iamBindingWithAndWithoutCondition
- TestAccStorageBucket_iamMemberWithAndWithoutCondition
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/7fd10201-1937-453c-96b1-19501055d8d9/build-log/replaying_test.log)

kautikdk reviewed (commented):
I believe we also need to handwrite documentation of the data source, google_storage_bucket_iam_policy.

Apart from that, LGTM!

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 125 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 125 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 118
Passed tests: 109
Skipped tests: 9
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccStorageBucket_iamBindingWithAndWithoutCondition
- TestAccStorageBucket_iamMemberWithAndWithoutCondition
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/c3525dfd-798d-4b08-bb49-bb6a863b026a/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 90 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 90 deletions(-))

gurusai-voleti commented:
> I believe we also need to handwrite documentation of the data source, google_storage_bucket_iam_policy.
> 
> Apart from that, LGTM!

@kautikdk added google_storage_bucket_iam_policy doc to handwritten

kautikdk commented on mmv1/third_party/terraform/services/storage/iam_storage_bucket_test.go:
```diff
@@ -0,0 +1,662 @@
+	})
+}
+
+func testAccStorageBucketIamPolicy_destroy() string {
```
nit: Individual config generator methods should be placed after all tests.

gurusai-voleti replied:
updated

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 90 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12944-old..auto-pr-12944) ( 5 files changed, 117 insertions(+), 90 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 118
Passed tests: 109
Skipped tests: 9
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storage</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccStorageBucket_iamBindingWithAndWithoutCondition
- TestAccStorageBucket_iamMemberWithAndWithoutCondition
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12944/artifacts/46a0236d-b9c6-4f9c-8d93-a56da3c04eb9/build-log/replaying_test.log)

gurusai-voleti commented:
@rileykarson  could you please review this PR

pavandagagoogle commented:
@rileykarson - Could you provide review on the same

rileykarson reviewed (approved):
LGTM

Sorry for the delay- I was hoping to find a better solution than dropping to handwritten here, but ended short on cycles to do so.
